cf_units/__init__.py

```diff
@@ -406,17 +406,20 @@ def _discard_microsecond(date):
     dates = np.asarray(date)
     shape = dates.shape
     dates = dates.ravel()
+    result = None
     # Create date objects of the same type returned by cftime.num2date()
     # (either datetime.datetime or cftime.datetime), discarding the
     # microseconds
-    dates = np.array(
-        [
-            d
-            and d.__class__(d.year, d.month, d.day, d.hour, d.minute, d.second)
-            for d in dates
-        ]
-    )
-    result = dates[0] if shape == () else dates.reshape(shape)
+    discard = [dt.replace(microsecond=0) for dt in dates if dt]
```

Comment:
OK I done a bit of digging.  I **think** the `d and` bit is so we can handle masked arrays, and went in at #68.  If we have a masked array of datetimes

```python
unit = cf_units.Unit('days since 1970-01-01')
dates = unit.num2date(np.ma.masked_array(range(5), mask=[0, 0, 1, 1, 0]))
```

If we just do

```python
[d.replace(microsecond=0) for d in dates]
```
we get an `AttributeError`.  If instead we do

```python
[d and d.replace(microsecond=0) for d in dates]
```
we get our masked array back.
```
[cftime.DatetimeGregorian(1970, 1, 1, 0, 0, 0, 0, has_year_zero=False),
 cftime.DatetimeGregorian(1970, 1, 2, 0, 0, 0, 0, has_year_zero=False),
 masked,
 masked,
 cftime.DatetimeGregorian(1970, 1, 5, 0, 0, 0, 0, has_year_zero=False)]
```

Which is different from using `if d`, as that reduces the size of the array

```python
[d.replace(microsecond=0) for d in dates if d]
```
```
[cftime.DatetimeGregorian(1970, 1, 1, 0, 0, 0, 0, has_year_zero=False),
 cftime.DatetimeGregorian(1970, 1, 2, 0, 0, 0, 0, has_year_zero=False),
 cftime.DatetimeGregorian(1970, 1, 5, 0, 0, 0, 0, has_year_zero=False)]
```

So I **think** we should still be using `dt and`.

Reply:
Cool. Thanks for the context... I'll change and extend the tests to cover this.